[inductor] Fix float64 constants in triton codegen

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #104725
* __->__ #104830

Fixes #101684

Before this change, we get a float constant in triton
```
tmp0 = 0.2
```
which in triton IR becomes a float32 value
```
%cst_0 = arith.constant dense<2.000000e-01> : tensor<2xf32>
```

After, we get a tensor with explicit type
```
tmp0 = tl.full([1], 0.2, tl.float64)
```
which does generate a float64 in the triton IR
```
%cst_0 = arith.constant dense<2.000000e-01> : tensor<2xf64>
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8

peterbell10 committed:
[inductor] Fix float64 constants in triton codegen

Fixes #101684

Before this change, we get a float constant in triton
```
tmp0 = 0.2
```
which in triton IR becomes a float32 value
```
%cst_0 = arith.constant dense<2.000000e-01> : tensor<2xf32>
```

After, we get a tensor with explicit type
```
tmp0 = tl.full([1], 0.2, tl.float64)
```
which does generate a float64 in the triton IR
```
%cst_0 = arith.constant dense<2.000000e-01> : tensor<2xf64>
```

[ghstack-poisoned]

lezcano reviewed (commented):
While this makes sense, wonder whether we'll end up having issues with kernels that have plenty of constants.

Also, what's triton's broadcasting rules? I hope this change does not type cast float tensors into double when multiplying them by a constant.

peterbell10 committed:
Update on "[inductor] Fix float64 constants in triton codegen"


Fixes #101684

Before this change, we get a float constant in triton
```
tmp0 = 0.2
```
which in triton IR becomes a float32 value
```
%cst_0 = arith.constant dense<2.000000e-01> : tensor<2xf32>
```

After, we get a tensor with explicit type
```
tmp0 = tl.full([1], 0.2, tl.float64)
```
which does generate a float64 in the triton IR
```
%cst_0 = arith.constant dense<2.000000e-01> : tensor<2xf64>
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

peterbell10 commented:
> wonder whether we'll end up having issues with kernels that have plenty of constants.

What sort of issues were you thinking of?

> Also, what's triton's broadcasting rules? I hope this change does not type cast float tensors into double when multiplying them by a constant.

In that case the `ops.constant` call has `torch.float32`. If there are cases that rely on `float64` actually secretly being `float32` then I would say that is a bug and it would be good to root it out.

lezcano commented:
I was thinking of register usage, as these take two registers rather than one, but yeah, these things should just be loaded into registers under demand, so that should be alright.

My point is that `torch.randn(3) * 0.1` in PyTorch results in a `float32` tensor, even if `0.1`is `float64` due to the type promotion rules we have, where scalars don't promote. I wonder whether this is still the case in triton.

lezcano commented on torch/_inductor/codegen/triton.py:
```diff
@@ -134,18 +134,21 @@ def to_dtype(x, dtype: torch.dtype):
-        else:
-            type_ = torch._prims_common.dtype_to_type(dtype)
-            return triton_constant(type_(value))
+            # tl.full is broken for uint8, remove once triton is fixed
```
link the relevant issue? Otherwise, how about doing the good old `value + tl.zeros`? Is that one also broken?

peterbell10 replied:
`tl.zeros` is implemented in terms of `tl.full` so is also broken.

peterbell10 commented:
> My point is that torch.randn(3) * 0.1 in PyTorch results in a float32 tensor, even if 0.1is float64 due to the type promotion rules we have, where scalars don't promote. I wonder whether this is still the case in triton.

`0.1 * tl.zeros([3], tl.float32)` results in a float32 tensor because `0.1` is a float32 already. If you do `tl.zeros([3], tl.float64) + tl.zeros[3], tl.float32)` then the result is `float64` as you would expect.

lezcano commented:
The case I'm concerned about is the lowering for something like `t + 3.0` where `t` is a `float32` tensor

The result of that operation in PyTorch is a tensor of dtype `float32`. Before this PR, we represent that operation in triton as
```python
tmp0 = tl.load(...)
tmp1 = 3.0
tmp2 = tmp0 + tmp1
```
Since `3.0` and `tmp0` are both `float32` in triton, then `tmp2` will be `float32`. Same as in PyTorch, even if the constant `tmp1` did not have the right dtype.

With this PR, we would do
````python
tmp1 = tl.full([1], 0.2, tl.float64)
```
Would in this case `tmp2` still be `float32`? In other words, does triton have semantics like NumPy, where tensors of `numel=1` do not participate in the calculation of the dtype of the result? If not, I'd be concerned about the perf hit of upcasting to `float64` in later ops in the graph.

lezcano commented on torch/_inductor/codegen/triton.py:
```diff
@@ -134,18 +134,21 @@ def to_dtype(x, dtype: torch.dtype):
+        # NOTE: We use a tensor here in order to get the correct type.
+        # Otherwise, e.g. float64 constants would be trunctated to float32.
+        ndim = V.kernel.triton_tensor_ndim()
+        shape = [1] * ndim
```
why do we need to do `[1] * ndim`? Isn't it enough to do `shape = [1]` and rely on broadcasting?

peterbell10 replied:
Yes broadcasting should be sufficient, but increasing ndim in triton IR is an extra operator, so the triton IR is cleaner if everything starts with the same ndim. Not a huge deal but also it's only a two-line change.

lezcano replied:
Could you leave a short comment?

lezcano commented on torch/_inductor/codegen/triton.py:
```diff
@@ -134,18 +134,21 @@ def to_dtype(x, dtype: torch.dtype):
+            tmp = cls.constant(value, torch.int16)
+            return cls.to_dtype(tmp, dtype)
+
+        # NOTE: We use a tensor here in order to get the correct type.
+        # Otherwise, e.g. float64 constants would be trunctated to float32.
```
We could default to the previous path for `int32` and `float32` constants, to get simpler codegen'd in those cases.

peterbell10 commented:
> The case I'm concerned about is the lowering for something like t + 3.0 where t is a float32 tensor

This is the IR for that code:
```python
def body(self, ops):
    get_index = self.get_index('index0')
    load = ops.load('arg0_1', get_index)
    constant = ops.constant(3.0, torch.float32)
    add = ops.add(load, constant)
    get_index_1 = self.get_index('index0')
    store = ops.store('buf0', get_index_1, add, None)
    return store
```

AsI said earlier, the `ops.constant` call has type `float32` here so there is no float64 usage in the kernel whatsoever. 

You can also see that if you force the constant to participate in type promotion by making it `torch.full((), 3.0 , dtype=torch.float64, device="cuda")` then the promotion to float64 is explicit in the inductor IR:

```python
def body(self, ops):
    get_index = self.get_index('index0')
    load = ops.load('arg0_1', get_index)
    to_dtype = ops.to_dtype(load, torch.float64)
    constant = ops.constant(3.0, torch.float64)
    add = ops.add(to_dtype, constant)
    get_index_1 = self.get_index('index0')
    store = ops.store('buf0', get_index_1, add, None)
    return store
```

This is why I say that triton's promotion rules shouldn't matter here.

peterbell10 commented:
And FWIW I don't think triton has shape-dependent type-promotion rules, that sounds like it would be a massive pain.

lezcano reviewed (approved):
So we treat the Python constants as float32 by default. I guess that makes sense.

Approving, but it'd be good to leave a couple comments as described in the previous reviews.

lezcano commented:
And also consider https://github.com/pytorch/pytorch/pull/104830#discussion_r1258143248

peterbell10 commented:
> So we treat the Python constants as float32 by default. I guess that makes sense.

If you're curious here is the function that decides the type of a constant
https://github.com/pytorch/pytorch/blob/af34123caf287df9697532eeeeb6906398928dfe/torch/_inductor/lowering.py#L334

peterbell10 committed:
Update on "[inductor] Fix float64 constants in triton codegen"


Fixes #101684

Before this change, we get a float constant in triton
```
tmp0 = 0.2
```
which in triton IR becomes a float32 value
```
%cst_0 = arith.constant dense<2.000000e-01> : tensor<2xf32>
```

After, we get a tensor with explicit type
```
tmp0 = tl.full([1], 0.2, tl.float64)
```
which does generate a float64 in the triton IR
```
%cst_0 = arith.constant dense<2.000000e-01> : tensor<2xf64>
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge

peterbell10 commented:
@pytorchbot merge
